doc/jemalloc.xml.in

```diff
@@ -181,17 +181,39 @@
       <title>Standard API</title>
 
       <para>The <function>malloc()</function> function allocates
-      <parameter>size</parameter> bytes of uninitialized memory.  The allocated
-      space is suitably aligned (after possible pointer coercion) for storage
-      of any type of object.</para>
+      <parameter>size</parameter> bytes of uninitialized memory. The allocated
+      space is suitably aligned (after possible pointer coercion) for storage of
+      any type of object. The meaning of "suitably aligned" is specified in the
+      C standard. Recall that, in C, the size of an object is a multiple of its
+      alignment (<code language="C">size >= alignment</code>). An allocation is
+      therefore always suited to hold any type of object if its alignment equals
+      the size of the object when the size is a power-of-two, or if the
```

Comment:
Is that true? Imagine an object of size 256 and alignment 256.

Reply:
> Is that true? Imagine an object of size 256 and alignment 256.

> An allocation is therefore always suited to hold any type of object if its alignment equals the size of the object when the size is a power-of-two, or if the alignment is computed by rounding the object size towards the previous power-of-two.

That's a claim that such types of allocations are always suited to hold any C object of any size / 
alignment pair, due to size >= alignment in C. 

An object of size 256 and alignment 256 would fit in an allocation of size 256 at an alignment boundary of 256. 

If you think this claim is false, could you explain why you think that ?

---

> The text below says basically "the C standard guarantees this only for small sizes", but does this sentence here mean that jemalloc guarantees it for all sizes?

The C standard says that `malloc` does not need to return allocations aligned to a boundary larger than `_Alignof(max_align_t)`, the largest _fundamental alignment_ that C objects are allowed to have. That is, if, e.g., `_Alignof(max_align_t) == 16`, `malloc` doesn't need to return an allocation suited for an object of size 256 and alignment 256. 

That's ok, because in C, objects of size 256 and alignment 256 do not have _fundamental alignment_. They are overaligned, and those are not "supported" by malloc.

Reply:
> If you think this claim is false, could you explain why you think that ?

I read this sentence as claiming that the returned allocation is suited to hold an object of a type with size and alignment 256. Is that really the case? Does jemalloc guarantee that 256-byte-sized allocations are 256-byte-aligned?

Reply:
Where does that sentence claim anything about the returned allocation ? 

It makes a claim about storage in general, not about anything returned by malloc.

Reply:
If "An allocation" here does not mean "the allocation provided by `jemalloc`, that should be made more explicit.

The more were are talking here, the less I know what this sentence says. ;)

Reply:
Reworded.

Reply:
Oh I think now I understand where you are getting. May I suggest the following?

> Recall that, in C, the size of an object is a positive multiple of its alignment (so in particular, <code language="C">size >= alignment</code>).
> As a consequence, for an allocation with alignment A to be suited to store an object of size N, A needs to be obtained by rounding N down to the nearest power of two (in particular, if N *is* a power of two, then N == A).

Reply:
Thanks, that will do. The next thing that must then be clear is that C standard only guarantees that malloc returns storage suitable for objects that have what C calls "fundamental alignment", that is, alignment <= _Alignof(max_align_t). If you have an object that doesn't satisfy this requirement, then malloc storage is not required to be suited for it.

Reply:
Yes, that is the immediate next sentence here.